azurelinuxagent/common/cgroupstelemetry.py

```diff
@@ -13,47 +13,188 @@
-    _tracked = {}
+    _tracked = []
+    _cgroup_metrics = {}
+    _rlock = threading.RLock()
 
     @staticmethod
-    def metrics_hierarchies():
-        return CGroupsTelemetry._hierarchies
+    def _get_metrics_list(metric):
+        return [metric.average(), metric.min(), metric.max(), metric.median(), metric.count(), metric.first_poll_time(),
+                metric.last_poll_time()]
+
+    @staticmethod
+    def _process_cgroup_metric(cgroup_metrics):
+        current_memory_usage, max_memory_levels, current_cpu_usage = cgroup_metrics.get_metrics()
+        processed_extension = {
+            "memory": {
+                "cur_mem": CGroupsTelemetry._get_metrics_list(current_memory_usage),
+                "max_mem": CGroupsTelemetry._get_metrics_list(max_memory_levels)
+            },
+            "cpu": {
+                "cur_cpu": CGroupsTelemetry._get_metrics_list(current_cpu_usage)
+            }
+        }
+        return processed_extension
 
     @staticmethod
     def track_cgroup(cgroup):
         """
-        Adds the given item to the list of tracked cgroups
+        Adds the given item to the dictionary of tracked cgroups
         """
-        pass  # TODO
+        # Making track thread-safe.
+        CGroupsTelemetry._rlock.acquire()
+        CGroupsTelemetry._tracked.append(cgroup)
+        CGroupsTelemetry._rlock.release()
+        logger.info("Started tracking new cgroup: {0}, path: {1}".format(cgroup.name, cgroup.path))
 
     @staticmethod
-    def is_tracked(name):
+    def is_tracked(name, controller):
         """
         Returns true if the given item is in the list of tracked items
+        O(n) operation. But limited to few cgroup objects we have.
         """
-        pass  # TODO
+        for cgroup in CGroupsTelemetry._tracked:
+            if name == cgroup.name and controller == cgroup.controller:
+                return True
+
+        return False
 
     @staticmethod
-    def stop_tracking(name):
+    def stop_tracking(cgroup):
         """
         Stop tracking the cgroups for the given name
         """
-        if CGroupsTelemetry.is_tracked(name):
-            del (CGroupsTelemetry._tracked[name])  # TODO - thread-safeness; review implementation
+        # Pop is atomic and thread-safe.
+        CGroupsTelemetry._tracked.remove(cgroup)
+        logger.info("Stopped tracking cgroup: {0}, path: {1}".format(cgroup.name, cgroup.path))
 
     @staticmethod
     def collect_all_tracked():
-        pass  # TODO
+        collected_metrics = {}
+        for name, cgroup_metrics in CGroupsTelemetry._cgroup_metrics.items():
+            collected_metrics[name] = CGroupsTelemetry._process_cgroup_metric(cgroup_metrics)
+            cgroup_metrics.clear()
+
+            if cgroup_metrics.marked_for_delete:
+                # Clear the _cgroup_metrics to make sure we don't send stale entries,
+                # only after sending its content first
+                CGroupsTelemetry._cgroup_metrics.pop(name, None)
```

Comment:
> only after sending its content first

Yes, that's what it should do... but we are clearing before sending

Reply:
Its metrics are captured in line 81 - `collected_metrics[name] = CGroupsTelemetry._process_cgroup_metric(cgroup_metrics)`, and we return `collected_metrics` from `collect_all_tracked()`